pkg/codegen/pcl/binder_resource.go

```diff
@@ -134,6 +136,95 @@ func AnnotateResourceInputs(node *Resource) {
 	}
 }
 
+func resolveUnionOfObjects(objectExpr *model.ObjectConsExpression, union *schema.UnionType) schema.Type {
+	var discriminatorValue string
+	for _, item := range objectExpr.Items {
+		if key, ok := item.Key.(*model.LiteralValueExpression); ok {
+			if key.Value.AsString() == union.Discriminator {
+				if value, ok := item.Value.(*model.LiteralValueExpression); ok {
+					discriminatorValue = value.Value.AsString()
+					break
+				}
+
+				if value, ok := item.Value.(*model.TemplateExpression); ok && len(value.Parts) == 1 {
+					if literalValue, ok := value.Parts[0].(*model.LiteralValueExpression); ok {
+						discriminatorValue = literalValue.Value.AsString()
+						break
+					}
+				}
+			}
+		}
+	}
+
+	if discriminatorValue == "" {
+		return union
+	}
+
+	if correspondingTypeToken, ok := union.Mapping[discriminatorValue]; ok {
+		for _, schemaType := range union.ElementTypes {
+			if schemaObjectType, ok := codegen.UnwrapType(schemaType).(*schema.ObjectType); ok {
+				parsedTypeToken, err := tokens.ParseTypeToken(correspondingTypeToken)
+				if err != nil {
+					continue
+				}
+
+				parsedObjectToken, err := tokens.ParseTypeToken(schemaObjectType.Token)
+				if err != nil {
+					continue
+				}
+
+				if string(parsedTypeToken.Name()) == string(parsedObjectToken.Name()) {
+					// found the corresponding object type
+					return schemaObjectType
+				}
+			}
+		}
+	}
+
+	return union
+}
+
+func resolveInputUnions(inputs map[string]model.Expression, inputProperties []*schema.Property) {
+	for _, property := range inputProperties {
+		if value, ok := inputs[property.Name]; ok {
+			switch schemaType := codegen.UnwrapType(property.Type).(type) {
+			case *schema.UnionType:
+				if objectExpr, ok := value.(*model.ObjectConsExpression); ok {
+					resolvedType := resolveUnionOfObjects(objectExpr, schemaType)
+					switch resolvedType := resolvedType.(type) {
+					case *schema.ObjectType:
+						// found the corresponding object type for this PCL expression, resolve nested unions if any
+						nestedInputs := map[string]model.Expression{}
+						for _, item := range objectExpr.Items {
+							if key, ok := literalExprValue(item.Key); ok && key.Type().Equals(cty.String) {
+								nestedInputs[key.AsString()] = item.Value
+							}
+						}
+
+						resolveInputUnions(nestedInputs, resolvedType.Properties)
+					}
+
+					property.Type = resolvedType
+				}
+			}
+		}
+	}
+}
+
+func (b *binder) rawResourceInputs(node *Resource) map[string]model.Expression {
+	inputs := map[string]model.Expression{}
+	scopes := newResourceScopes(b.root, node, nil, nil)
+	block, _ := model.BindBlock(node.syntax, scopes, b.tokens, b.options.modelOptions()...)
+	for _, item := range block.Body.Items {
+		switch item := item.(type) {
+		case *model.Attribute:
+			inputs[item.Name] = item.Value
+		}
+	}
+
+	return inputs
+}
```

Comment:
Can we document these functions -- why they exist, what they're doing, etc.? The PR description is excellent and thorough -- would be a shame to lose that context in the code itself.

Reply:
I think it could be beneficial to document the other two functions here, and potentially give a motivating example/really explain _why_ it's important that we do this (just like the excellent PR description!) I feel like the "why" and "show me what you mean" (e.g. with an example) is going to be key for anyone approaching this code later on, should we have to modify/extend it etc.

Reply:
I added more docs. Hopefully this is sufficient to figure out what the code is doing 🙏 can you have another look?